src/python_testing/mdns_discovery/mdns_discovery.py

```diff
@@ -15,385 +15,549 @@
-                            purposes. Defaults to False.
+            service_types (Optional[List[str]]): Specific service types to discover
+                (e.g., ['_matterc._udp.local.']).
+            all_services (bool): If True, discovers all available mDNS services.
+            discovery_timeout_sec (float): Maximum time in seconds to wait for service
+                announcements during the discovery phase. Defaults to DISCOVERY_TIMEOUT_SEC (15 sec).
+            query_service (bool): If True, queries the full service info for each
+             discovered service name. Defaults to False.
+            query_timeout_sec (float): Maximum time in seconds to wait for the record
+                query phase. Only applies when `query_service` is True. Defaults to QUERY_TIMEOUT_SEC (10 sec).
+            append_results (bool): If True, appends results to `self._discovered_services`
+                without clearing previous entries. If False, clears the dictionary before
+                storing new results. Only applies when `query_service` is True.
+            log_output (bool, optional): If True, logs the discovered services to the
+                console. Defaults to False.
+
+        Raises:
+            ValueError: If both `all_services` and `service_types` are provided.
+            ValueError: If `service_types` is provided but is an empty list.
+            ValueError: If `query_timeout_sec` is set when `query_service` is False.
+            ValueError: If `append_results` is True when `query_service` is False.
 
         Returns:
-            None: This method does not return any value.
-
-        Note:
-            The discovery duration may need to be adjusted based on network conditions and expected response times for service
-            announcements. The method leverages an asyncio event to manage asynchronous waiting and cancellation based on discovery
-            success or timeout.
+            None: This method does not return any value.`
         """
-        self._event.clear()
+        if all_services and service_types:
+            raise ValueError("Specify either 'all_services' or 'service_types', not both.")
 
-        if all_services:
-            self._service_types = list(set(await AsyncZeroconfServiceTypes.async_find()))
+        if service_types is not None and not service_types:
+            raise ValueError("`service_types` cannot be an empty list.")
 
-        logger.info(f"Browsing for MDNS service(s) of type: {self._service_types}")
+        if not query_service and query_timeout_sec != QUERY_TIMEOUT_SEC:
+            raise ValueError("`query_timeout_sec` is only valid when `query_service=True`.")
 
-        aiobrowser = AsyncServiceBrowser(zeroconf=self._zc,
-                                         type_=self._service_types,
-                                         handlers=[self._on_service_state_change]
-                                         )
+        if not query_service and append_results:
```

Comment:
Would it be very disruptive to have a type that does not allow invalid inputs, so that this kind of checks are not needed?

I imagine to have something like `service_query: Optional[ServiceQuery] = None`

and have

```py
@dataclass
class ServiceQuery:
    timeout_sec = QUERY_TIMEOUT_SEC
    append_results = False
```

And then it is impossible to have conflicting values (either you query services or you do not).